terraform/context_plan_test.go

```diff
@@ -4992,8 +4993,10 @@ func TestContext2Plan_createBeforeDestroy_depends_datasource(t *testing.T) {
 		}
 	}
 	p.ReadDataSourceFn = func(req providers.ReadDataSourceRequest) providers.ReadDataSourceResponse {
+		cfg := req.Config.AsValueMap()
```

Comment:
Yes, we will need to call this out. I'm sure we'll find someone why relies on using stale data sources, but since we want to eventually move this completely into planning, it's something we need to deal with.

Reply:
Having had some time away to reflect on this a bit more, I'm feeling like this half-way position where we read data resources during refresh _and_ during plan is likely to be confusing, because it will still be subject to the usual quirks/bugs where data sources get resolved against the "wrong" values (the ones in the prior state, rather than the ones in the config) but it will also no longer be disableable with `-refresh=false` and so the user model here is kinda odd.

I realize I'm taking a different posture here than my initial reaction to this changeset, but out of curiosity: do you have a sense of the size of the delta from this PR to having data sources not update during refresh at all? I was previously feeling nervous about squeezing that in so late before 0.13, but the effect of _this_ change feels potentially heavy too and so now I'm wondering if we shouldn't just go all-in on the new behavior and get some more benefit to justify the risk. What do you think? :thinking: 

(I'm also now considering that this would be released at the same time as an initial iteration of #15419, which wasn't true when we originally discussed this, and so that capability could help mitigate the fact that `terraform refresh` would no longer work to get data sources updated because folks could use `terraform apply` with no other changes to get there instead.)

Reply:
The benefit with this PR is that while refresh can be using "old" values for data source config, plan will check again if the data source needs to be read, and update the data source there if needed. While overall it is still not optimal, it should reduce the confusing instances to only those where either the evaluation fails entirely (which is the same situation we have now requiring `-refresh=false`), or when providers require a change in the data source via a dependency (which again we have today).

Overall there shouldn't be much more work to _remove_ data source reevaluation from a separate refresh cycle, but that's not what we need in the end. The bulk of the remaining work is removing the separate refresh cycle entirely, which requires the merging of the refresh and plan walks for all resources at once, which did not look trivial (I only briefly ventured down that path). I also toyed with both not having data sources refresh, and having data sources refresh from state, but that always fails with the cases around using data sources for provider configuration, especially when the data source is involved in authentication.

Reply:
Hmm indeed... the abuse of data sources for issuing credentials is definitely a wrench in the works here. Reading data sources twice during a plan is not great for that situation either, because it will presumably then issue _two_ sets of credentials, though that's better than it failing altogether.

This particular data source misuse is going to be problematic for #15419 too (`terraform plan` will never succeed without generating a new plan to be applied) so I expect we're going to need to address it somehow either way... :confounded:

Reply:
The good news is that the plan-time read is conditional, so it only get's issued twice if the configuration has visible changes (or if the data is doing something incorrectly, which also isn't out of the question with the relaxed legacy API constraints).